frame/name-service/src/commit_reveal.rs

```diff
@@ -0,0 +1,183 @@
+use sp_std::prelude::*;
+
+impl<T: Config> Pallet<T> {
+	/// Get the commitment hash from the raw name and secret.
+	pub fn commitment_hash(name: &[u8], secret: u64) -> CommitmentHash {
+		sp_io::hashing::blake2_256(&(name, secret).encode())
+	}
+
+	/// Get the name hash from raw bytes.
+	pub fn name_hash(name: &[u8]) -> NameHash {
+		sp_io::hashing::blake2_256(name)
+	}
+
+	/// Returns a commitment by hash if it exists.
+	pub fn get_commitment(
+		commitment_hash: CommitmentHash,
+	) -> Result<CommitmentOf<T>, DispatchError> {
+		Commitments::<T>::get(commitment_hash).ok_or(Error::<T>::CommitmentNotFound.into())
+	}
+
+	/// Checks whether a commitment has passed the minimum commitment period.
+	pub fn is_commitment_valid(
+		commitment: &CommitmentOf<T>,
+		block_number: &T::BlockNumber,
+	) -> bool {
+		&commitment.when.saturating_add(T::MinCommitmentAge::get()) < block_number
+	}
+
+	/// Checks whether a commitment has passed the commitment expiry time.
+	pub fn is_commitment_expired(
+		commitment: &CommitmentOf<T>,
+		block_number: &T::BlockNumber,
+	) -> bool {
+		&commitment.when.saturating_add(T::MaxCommitmentAge::get()) < block_number
+	}
+
+	pub fn do_commit(
+		depositor: T::AccountId,
+		owner: T::AccountId,
+		commitment_hash: CommitmentHash,
+	) -> DispatchResult {
+		ensure!(!Commitments::<T>::contains_key(commitment_hash), Error::<T>::CommitmentExists);
+
+		let block_number = frame_system::Pallet::<T>::block_number();
+		let deposit = T::CommitmentDeposit::get();
+
+		T::Currency::reserve(&depositor, deposit)?;
+
+		let commitment = Commitment {
+			owner: owner.clone(),
+			when: block_number,
+			depositor: depositor.clone(),
+			deposit,
+		};
+
+		Commitments::<T>::insert(commitment_hash, commitment);
+		Self::deposit_event(Event::<T>::Committed { depositor, owner, hash: commitment_hash });
+		Ok(())
+	}
+
+	pub fn do_reveal(
+		fee_payer: T::AccountId,
+		name: Vec<u8>,
+		secret: u64,
+		length: T::BlockNumber,
+	) -> DispatchResult {
+		ensure!(name.len() <= T::MaxNameLength::get() as usize, Error::<T>::NameTooLong);
+
+		let commitment_hash = Self::commitment_hash(&name, secret);
+		let commitment = Self::get_commitment(commitment_hash)?;
+
+		let block_number = frame_system::Pallet::<T>::block_number();
+
+		ensure!(
+			Self::is_commitment_valid(&commitment, &block_number),
+			Error::<T>::TooEarlyToReveal
+		);
+
+		let name_hash = sp_io::hashing::blake2_256(&name);
+
+		ensure!(Self::get_registration(name_hash).is_err(), Error::<T>::RegistrationExists);
+
+		let fee = Self::registration_fee(name.clone(), length);
+
+		let imbalance = T::Currency::withdraw(
+			&fee_payer,
+			fee,
+			WithdrawReasons::FEE,
+			ExistenceRequirement::KeepAlive,
+		)?;
+
+		T::RegistrationFeeHandler::on_unbalanced(imbalance);
+
+		let expiry = block_number.saturating_add(length);
+
+		// TODO AUDIT WARNING: Can return an error after `withdraw` happens. Needs to be
+		// transactional or refactored.
+		Self::do_register(
```

Comment:
I guess this could just be moved up before calling `T::RegistrationFeeHandler::on_unbalanced`? Then there are no failures after this is executed.